lib/api/src/js/module.rs

```diff
@@ -274,10 +268,71 @@ impl Module {
+        let imports_object = js_sys::Object::new();
+        let mut import_externs: Vec<Extern> = vec![];
+        for import_type in self.imports() {
+            let resolved_import = imports.get_export(import_type.module(), import_type.name());
+            #[allow(unused_variables)]
+            if let wasmer_types::ExternType::Memory(mem_ty) = import_type.ty() {
+                if resolved_import.is_some() {
+                    #[cfg(feature = "tracing")]
+                    debug!("imported shared memory {:?}", &mem_ty);
+                } else {
+                    #[cfg(feature = "tracing")]
+                    warn!(
+                        "Error while importing {0:?}.{1:?}: memory. Expected {2:?}",
+                        import_type.module(),
+                        import_type.name(),
+                        import_type.ty(),
+                    );
+                }
+            }
+            #[allow(unused_unsafe)]
+            unsafe {
+                if let Some(import) = resolved_import {
+                    let val = js_sys::Reflect::get(&imports_object, &import_type.module().into())?;
+                    if !val.is_undefined() {
+                        // If the namespace is already set
+                        js_sys::Reflect::set(
+                            &val,
+                            &import_type.name().into(),
+                            &import.as_jsvalue(&store.as_store_ref()),
+                        )?;
+                    } else {
+                        // If the namespace doesn't exist
+                        let import_namespace = js_sys::Object::new();
+                        js_sys::Reflect::set(
+                            &import_namespace,
+                            &import_type.name().into(),
+                            &import.as_jsvalue(&store.as_store_ref()),
+                        )?;
+                        js_sys::Reflect::set(
+                            &imports_object,
+                            &import_type.module().into(),
+                            &import_namespace.into(),
+                        )?;
+                    }
+                    import_externs.push(import);
+                } else {
+                    #[cfg(feature = "tracing")]
+                    warn!(
+                        "import not found {}:{}",
+                        import_type.module(),
+                        import_type.name()
+                    );
+                }
+            }
+            // in case the import is not found, the JS Wasm VM will handle
+            // the error for us, so we don't need to handle it
+        }
+        Ok((
+            crate::StoreHandle::new(
+                store.as_store_mut().objects_mut(),
+                WebAssembly::Instance::new(&self.module, &imports_object)
+                    .map_err(|e: JsValue| -> RuntimeError { e.into() })?,
+            ),
+            import_externs,
+        ))
```

Comment:
I'm not sure this changes are really needed. Why do we need this?

Reply:
We need to ask @john-sharratt .

Reply:
The imported/exported memory does not have the correct properties (incorrect size and shared flag) hence when using shared memory its failing - the only way to fix it is to resolve the import and use the correct memory properties. this regression issue was only found in WASIX on the browser as the other areas don't mind that they don't match up

Reply:
Can you add a test/example that reproduces this? I think we can simplify but without a base test/example that we can test against it will be hard

Reply:
`sharrattj/dash` should be able to reproduce this but the problem is that's going to be quite hard now that the merge is split between two as it won't work without the second commit.

basically if dash can run (including with coreutils commands) without this piece of code then we can delete it

Reply:
Code now has a TODO to refactor this.